services/mpa/mpa.proto

```diff
@@ -0,0 +1,120 @@
+  // are stored in-memory and older requests may be cleared automatically.
+  // This call is idempotent - requests from the same user with the same
+  // contents will return the same id.
+  //
+  // The user for the request and any justification is implicitly passed in
+  // via inspecting the peer of the RPC or via gRPC metadata.
+  rpc Store(StoreRequest) returns (StoreResponse) {}
+  // Approve a previously stored request. A request can have multiple approvals
+  // and can be approved by anybody but the original user that stored the request.
+  //
+  // The user for the request is implicitly passed in via inspecting the
+  // peer of the RPC or via gRPC metadata.
+  rpc Approve(ApproveRequest) returns (ApproveResponse) {}
+  // Block until at least one approval has been granted. This is used
+  // as an optimization to avoid needing to poll for MPA approval.
+  rpc WaitForApproval(WaitForApprovalRequest) returns (WaitForApprovalResponse) {}
+  // List available requests.
+  rpc List(ListRequest) returns (ListResponse) {}
+  // Get a request and all approvals associated with it.
+  rpc Get(GetRequest) returns (GetResponse) {}
+  // Clear a stored request.
+  //
+  // This is typically unnecessary due to how requests are stored in memory
+  // and forgotten after a process restart.
+  rpc Clear(ClearRequest) returns (ClearResponse) {}
+}
+
+message Action {
+  // The user that created the request.
+  string user = 1;
+  // User-supplied information on why the request is being made.
+  string justification = 2;
+  // The GRPC method name, as '/Package.Service/Method'
+  string method = 3;
+  // The request protocol buffer.
+  google.protobuf.Any message = 4;
+}
+
+message Principal {
+  // The principal identifier (e.g. a username or service role)
+  string id = 1;
+  // Auxiliary groups associated with this principal.
+  repeated string groups = 2;
+}
+
+
+message StoreRequest {
+  // The GRPC method name, as '/Package.Service/Method'
+  string method = 1;
+  // The request protocol buffer. 
+  google.protobuf.Any message = 2;
+}
+
+message StoreResponse {
+  string id = 1;
+  Action action = 2;
+  // All approvers of the request. Storing is idempotent, so
+  // approvers may be non-empty if we're storing a previously
+  // approved command.
+  repeated Principal approver = 3;
+}
+
+message ApproveRequest {
+  // Approve takes an action instead of an ID to improve auditability
+  // and allow richer authorization logic.
+  Action action = 1;
+}
+message ApproveResponse {}
+
+message WaitForApprovalRequest { string id = 1; }
+message WaitForApprovalResponse {}
+
+message ListRequest {}
+message ListResponse {
+  message Item {
+    Action action = 1;  
+    repeated Principal approver = 2;
+    string id = 3;
+  }
+  repeated Item item = 1;
+}
+
+message GetRequest { string id = 1; }
+
+message GetResponse {
+  Action action = 1;
+  
+  // All approvers of the request.
+  repeated Principal approver = 2;
+}
```

Comment:
This will sometimes contain sensitive data.  The (reasonable) advice from the README.md "It's highly recommended to additionally allow `/Mpa.Mpa/List`, `/Mpa.Mpa/Get`" runs the risk that you'll expose that sensitive data to a wide audience.  Complex policies could be written, but perhaps a simpler approach is for `StoreRequest` to allow the user to send a read ACL, which `Get` could enforce?  WDYT?

Reply:
This part's a bit tricky. Get intentionally has nothing other than an ID in the request so rego policies can't do useful access control based on the request contents. I don't want to stick a read acl in StoreRequest for a couple reasons.

1. Users would have to specify more than `-mpa` when creating an MPA request
2. The access control is different than anything else we have and we'd either have a non-rego way of doing it or we'd have to have a weird rpcauth hook to allow doing it through rego

I've tweaked the wording a bit to say "let approvers do it" instead of "let anyone do it". In practice I think matching against a combination of peer identity and host information can give reasonable protection on who sees the data..